Detect serving cert changes in operator instead of hiveadmission.

dgoodwin opened the pull request:
Trying an alternate approach for detecting changes to the serving cert
by monitoring the secret in the hive-operator, and on any change we
reconcile. Reconcile will then hash the secret data and store in an
annotation on the pod template in the hiveadmission deployment. Any
change to this value will result in a new rollout.

dgoodwin commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -201,24 +208,26 @@ var _ reconcile.Reconciler = &ReconcileHiveConfig{}
+	managedConfigCMLister             corev1listers.ConfigMapLister
+	ctrlr                             controller.Controller
+	servingCertSecretWatchEstablished bool
+	mgr                               manager.Manager
```
Last three fields are the new ones.

dgoodwin commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -246,13 +255,63 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
+			UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+				log.Info("eventHandler UpdateFunc")
+				q.Add(reconcile.Request{NamespacedName: types.NamespacedName{Name: hiveConfigName}})
+			},
```
Tested and this does appear to work even though it's configured later than most watches are. It does not fire for secrets in other namespaces, and it fires for secrets in this namespace but filters out via the predicate.

staebler commented on pkg/operator/hive/hiveadmission.go:
```diff
@@ -142,8 +140,28 @@ func (r *ReconcileHiveConfig) deployHiveAdmission(hLog log.FieldLogger, h *resou
 			return err
 		}
+	} else {
+		// Set the serving cert CA secret hash as an annotation on the pod template to force a rollout in the event it changes:
+		servingCertSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: hiveNSName, Name: hiveAdmissionServingCertSecretName}, servingCertSecret); err != nil {
+			hLog.WithError(err).WithField("secretName", hiveAdmissionServingCertSecretName).Log(
+				controllerutils.LogLevel(err), "error getting serving cert secret")
+		}
+		hLog.Info("Running on OpenShift 4, hashing serving cert secret onto a hiveadmission deployment annotation")
+		certSecretHash := computeSecretDataHash(servingCertSecret.Data)
+		if hiveAdmDeployment.Spec.Template.Annotations == nil {
+			hiveAdmDeployment.Spec.Template.Annotations = map[string]string{}
+		}
+		hiveAdmDeployment.Spec.Template.Annotations[servingCertSecretHashAnnotation] = certSecretHash
```
Does this need to be gated on whether Hive is running on OpenShift 4? Even when Hive is running on OpenShift 3.11 or another kube flavor, the same serving cert is in use.

dgoodwin replied:
Good catch, I did not realize that.

dgoodwin commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -245,13 +254,63 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
+				return e.Meta.GetName() == hiveAdmissionServingCertSecretName
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// WARNING: can't check namespace for same reason as above, we don't have a HiveConfig
```
My bad, this was leftover from before I tried moving it into the reconcile loop itself.

dgoodwin commented:
Updated. Certs function is restored.

For the deletion of HiveConfig, when reconcile and find no HiveConfig "hive", the var is now cleared. The watch lives on, I can't see any way to delete it, so if you then go edit hiveadmission-serving-cert we will reconcile, lookup HiveConfig, log that it's deleted and return.

If you reinstall to the same ns the watch is recreated but it does not appear to trigger duplicate reconciles.

If you reinstall to *another* namespace, you now have two watches for the secrets in either ns. Editing either will trigger a normal reconcile. I think we can live with this as it's harmless, extremely unlikely to occur, and fixed as soon as you restart the pod.

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -245,13 +255,19 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
 
+	hiveNSName := getHiveNamespace(instance)
+
+	err = r.establishSecretWatch(hLog, hiveNSName)
```
`if err := ...; err != nil {`

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -337,6 +353,54 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
+		hLog.WithField("namespace", hiveNSName).Info("establishing watch on secrets in hive namespace")
+
+		// Create an informer that only listens to events in the OpenShift managed namespace
+		kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(r.kubeClient, 10*time.Minute,
```
10 minutes seems like a very low resync period. Does that controller-runtime end up setting a larger resync period? Or will the hive-operator be resync HiveConfig every 10 minutes?

dgoodwin replied:
I picked this up from the configmap watch for kube-apiserver-aggregator-client-ca above. I can bump them both to 30.

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -337,6 +353,54 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
+		kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(r.kubeClient, 10*time.Minute,
+			kubeinformers.WithNamespace(hiveNSName))
+		secretsInformer := kubeInformerFactory.Core().V1().Secrets().Informer()
+		err := r.mgr.Add(&informerRunnable{informer: secretsInformer})
```
`if err := ...; err != nil {`

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -337,6 +353,54 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
+		// force a deployment rollout.
+		err = r.ctrlr.Watch(&source.Informer{Informer: secretsInformer}, handler.Funcs{
+			CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+				log.Debug("eventHandler CreateFunc")
```
Should this use `hLog` instead of `log`? It would also be nice if the message were a bit more descriptive--at least mentioning that it is the event handler for the secret watcher.

dgoodwin replied:
Yup another remnant from when it was outside the reconciler. Updated.

dgoodwin commented:
Updated.

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -231,6 +243,7 @@ func (r *ReconcileHiveConfig) Reconcile(request reconcile.Request) (reconcile.Re
 			// Object not found, return.  Created objects are automatically garbage collected.
 			// For additional cleanup logic use finalizers.
 			hLog.Debug("HiveConfig not found, deleted?")
+			r.servingCertSecretWatchEstablished = false
```
This should only be cleared if the `HiveConfig` deleted was named `hive`. Maybe we should move the name check to before the `Get`.

dgoodwin replied:
Moved above and changed to not use instance.Name.

staebler commented on pkg/operator/hive/hive_controller.go:
```diff
@@ -200,29 +210,40 @@ var _ reconcile.Reconciler = &ReconcileHiveConfig{}
+
+	// We only support one HiveConfig per cluster, and it must be called "hive". This prevents installing
+	// Hive more than once in the cluster.
+	if request.NamespacedName.Name != hiveConfigName {
+		hLog.WithField("hiveConfig", request.NamespacedName.Name).Warn(
```
We can leave as is, but you do not need to include `NamespacedName`.

```suggestion
	if request.Name != hiveConfigName {
		hLog.WithField("hiveConfig", request.Name).Warn(
```

dgoodwin commented:
My long run test was successful, pods rotated and API stayed up without error.
